physical/raft/fsm.go

```diff
@@ -147,64 +164,31 @@ func NewFSM(conf map[string]string, logger log.Logger) (*FSM, error) {
 				return err
 			}
 
-			latestConfig.Store(&latest)
+			f.latestConfig.Store(&latest)
 		}
 		return nil
 	})
 	if err != nil {
-		return nil, err
-	}
-
-	storeLatestState := true
-	if _, ok := conf["doNotStoreLatestState"]; ok {
-		storeLatestState = false
-	}
-
-	f := &FSM{
-		path:   conf["path"],
-		logger: logger,
-
-		db:               boltDB,
-		latestTerm:       latestTerm,
-		latestIndex:      latestIndex,
-		latestConfig:     latestConfig,
-		storeLatestState: storeLatestState,
+		return err
 	}
 
-	f.chunker = raftchunking.NewChunkingBatchingFSM(f, &FSMChunkStorage{
-		f:   f,
-		ctx: context.Background(),
-	})
-
-	return f, nil
+	f.db = boltDB
+	return nil
 }
 
-// LatestState returns the latest index and configuration values we have seen on
-// this FSM.
-func (f *FSM) LatestState() (*IndexValue, *ConfigurationValue) {
-	return &IndexValue{
-		Term:  atomic.LoadUint64(f.latestTerm),
-		Index: atomic.LoadUint64(f.latestIndex),
-	}, f.latestConfig.Load().(*ConfigurationValue)
-}
+func (f *FSM) Close() error {
+	f.l.RLock()
```

Comment:
Should this be a write lock?

Reply:
No, it just needs read access to the pointer. See: https://godoc.org/github.com/etcd-io/bbolt#DB.Close

Reply:
It looks like `Close()` [does its own locking](https://github.com/etcd-io/bbolt/blob/master/db.go#L472). Just for my own edification, what does the additional locking here get us?

Reply:
Since we overwrite the db pointer on snapshot restore we need to guard all read access to the `f.db` pointer.